### PR TITLE
Width-3 Poseidon chip

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

File: src/circuit/gadget/poseidon/pow5t3.rs [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+/// Configuration for an [`Pow5T3Chip`].
+#[derive(Clone, Debug)]
+pub struct Pow5T3Config<F: FieldExt> {
+    state: [Column<Advice>; WIDTH],
+    state_permutation: Permutation,
+    partial_sbox: Column<Advice>,
+    rc_a: [Column<Fixed>; WIDTH],
+    rc_b: [Column<Fixed>; WIDTH],
+    s_full: Column<Fixed>,
+    s_partial: Column<Fixed>,
```
Comment (str4d) [inline src/circuit/gadget/poseidon/pow5t3.rs:35]: Change these to `Selector`s.

File: src/circuit/gadget/poseidon/pow5t3.rs [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+
+impl<F: FieldExt> Pow5T3Chip<F> {
+    /// Configures this chip for use in a circuit.
+    pub fn configure<S: Spec<F>>(
+        meta: &mut ConstraintSystem<F>,
+        spec: S,
+        state: [Column<Advice>; WIDTH],
+    ) -> Pow5T3Config<F> {
+        // Generate constants for the Poseidon permutation.
+        // This gadget requires R_F and R_P to be even.
```
Comment (str4d) [inline src/circuit/gadget/poseidon/pow5t3.rs:59]:
```suggestion
        assert_eq!(S::width(), 3);
        // This gadget requires R_F and R_P to be even.
```

Comment (str4d) [reply src/circuit/gadget/poseidon/pow5t3.rs:59]: Now that we're using const generics, this is unnecessary 😄

**Review by therealyingtong — APPROVED** [review]

**Comment by str4d** [comment]
CI failures are due to a non-conflicting merge conflict (feature flags added in separate PRs). I'll rebase this once #97 merges, both to fix the conflict, and so we can use the improvements to the `halo2` API and dev tooling.

**Comment by str4d** [comment]
Rebased to fix the merge conflict, and migrate the chip to the new `halo2` API. The gates have now been merged together to reflect each gate having multiple polynomial constraints.

**Review by ebfull — APPROVED** [review]

**Review by therealyingtong — APPROVED** [review]
utACK 91db490e202dbd19d96e0ae5519a2445e614534b

**Pull request merged**